docs/source/notes/cuda.rst

```diff
@@ -44,6 +44,68 @@ Below you can find a small example showcasing this::
 Best practices
 --------------
 
+Device-agnostic code
+^^^^^^^^^^^^^^^^^^^^
+
+Due to the structure of PyTorch, you may need to explicitly write
+device-agnostic (CPU or GPU) code; an example may be creating a new tensor as
+the initial hidden state of a recurrent neural network. 
+
+The first step is to determine whether the GPU should be used or not. A common
+pattern is to use Python's `argparse` module to read in user arguments, and
+have a flag that can be used to disable CUDA, in combination with
+`torch.cuda.is_available()`. In the following, `args.cuda` results in a flag
+that can be used to cast tensors and modules to CUDA if desired::
+
+    import argparse
+    import torch
+
+    parser = argparse.ArgumentParser(description='PyTorch Example')
+    parser.add_argument('--disable-cuda', action='store_true',
+                        help='Disable CUDA')
+    args = parser.parse_args()
+    args.cuda = not args.disable_cuda and torch.cuda.is_available()
+
+If modules or tensors need to be sent to the GPU, `args.cuda` can be used as
+follows::
+
+    x = torch.Tensor(8, 42)
+    net = Network()
+    if args.cuda:
+      x = x.cuda()
+      net.cuda()
+
+To avoid the if statement for tensors, an alternative is to have a default
+datatype defined, and cast all tensors using that. An example when using a
+dataloader would be as follows::
+
+    dtype = torch.cuda.FloatTensor
+    for i, x in enumerate(train_loader):
+        x = Variable(x.type(dtype))
+
+If you have a tensor and would like to create a new tensor of the same type,
+then you can use the `.new()` function, which acts the same as a normal tensor
+constructor. This is the recommended practice when creating modules in which
+new tensors/variables need to be created internally during the forward pass::
```

Comment:
This should mention that it acts as a regular constructor, except that it keeps the device, even if it’s different than the one of your context.

Reply:
Will do - what do you exactly mean by "context" though?

Reply:
Your code always runs in a "context" of a currently selected device. You can change it with context managers: `with torch.cuda.device(5):`

Reply:
Good point - I rarely have access to more than 1 GPU and just use DataParallel if I do. Added context management in now, let me know if there's anything else worth mentioning.